File: src/block_encoding.py
#!/usr/bin/env python3
# Author: Armit
# Create Time: 2024/05/20

from typing import List
import numpy as np
from numpy import ndarray
from pyqpanda import CPUQVM, QVec
from pyqpanda import QCircuit, H, I, X, Y, Z, RY, RZ, CNOT, SWAP, BARRIER
from pyqpanda import QProg, draw_qprog
from pyqpanda import matrix_decompose_hamiltonian, amplitude_encode, Encode
from pyqpanda import matrix_decompose, DecompositionMode

DEBUG = False
QCIR_SAVE = False
QCIR_LEN = 200

PauliI = np.asarray([
  [1, 0],
  [0, 1],
])
PauliX = np.asarray([
  [0, 1],
  [1, 0],
])
PauliY = np.asarray([
  [0, -1j],
  [1j, 0],
])
PauliZ = np.asarray([
  [1, 0],
  [0, -1],
])

# predefined matrices for test
A_complex = np.asarray([
  [0.1+0.2j, -0.3],
  [0, -0.4j],
])
A_real = np.asarray([
  [0.1, 0],
  [-0.2, 0.3],
])
# PauliY is not supported by pyQPanda :(
A_LCU = 0.1 * PauliI + 0.2 * PauliX + 0.4 * PauliZ


def show_circuit_matrix_UA(A:ndarray, qvm:CPUQVM, qv:QVec, qcir:QCircuit, lmbd:float):
  n_qubit_ex = len(qv)
  N_ex = 2 ** n_qubit_ex
  U_A = np.zeros([N_ex, N_ex], dtype=np.complex64)
  for i in range(N_ex):
    r = i
    cond = QCircuit()
    for j in range(n_qubit_ex):
      if r & 1: cond << X(qv[j])
      r >>= 1
    qprog = QProg() << cond << BARRIER(qv) << qcir
    qvm.directly_run(qprog)
    qs = qvm.get_qstate()
    U_A[:, i] = np.asarray(qs)

  if np.abs(U_A.imag).max() < 1e-5:
    U_A = U_A.real
  N = A.shape[0]
  U_A = U_A[:N, :N]

  print('A:')
  print(A.round(4))
  print('U_A:')
  print(U_A.round(4))
  print('U_A * lmbd:')
  print((U_A * lmbd).round(4))


def block_encoding_QSVT(A:ndarray) -> ndarray:
  def sqrt_matrix(A:ndarray):
    evs, vecs = np.linalg.eigh(A)
    evs = np.real(evs)
    evs = np.where(evs > 0.0, evs, 0.0)
    evs = evs.astype(vecs.dtype)
    return vecs @ np.diag(np.sqrt(evs)) @ vecs.conj().T

  N = A.shape[0]
  At = A.conj().T
  nq = int(np.log2(A.shape[0]))
  I = np.eye(2**nq)
  U_A = np.zeros_like(np.eye(2**(nq+1)), dtype=np.complex64)
  U_A[:N, :N] = A
  U_A[:N, N:] = sqrt_matrix(I - A @ At)
  U_A[N:, :N] = sqrt_matrix(I - At @ A)
  U_A[N:, N:] = -At

  qvm = CPUQVM()
  qvm.init_qvm()
  qv = qvm.qAlloc_many(nq + 1)
  qcir = matrix_decompose(qv, U_A, DecompositionMode.QSDecomposition)

  print(draw_qprog(qcir, line_length=QCIR_LEN))
  if QCIR_SAVE: draw_qprog(qcir, line_length=QCIR_LEN, output='pic', filename='qcir-QSVT.png')
  if DEBUG: show_circuit_matrix_UA(A, qvm, qv, qcir, lmbd=1)

print('[block_encoding_QSVT]')
block_encoding_QSVT(A_complex)


def block_encoding_LCU(A:ndarray) -> ndarray:
  ops = matrix_decompose_hamiltonian(A)
  if DEBUG: print('LCU:', ops)
  coeffs = []
  terms = []
  for what, coeff in ops.data():
    coeffs.append(coeff.real)
    terms.append(what[-1][0] if what[-1] else 'I')
  if DEBUG:
    print('coeffs:', coeffs)
    print('terms:', terms)

  n_qubit = int(np.ceil(np.log2(A.shape[0])))
  n_unitary = len(coeffs)
  n_ancilla = int(np.ceil(np.log2(n_unitary)))
  n_qubit_ex = n_qubit + n_ancilla
  if DEBUG:
    print('n_unitary:', n_unitary)
    print('n_ancilla:', n_ancilla)
    print('n_qubit_ex:', n_qubit_ex)

  qvm = CPUQVM()
  qvm.init_qvm()
  # |a,w>
  qv_w = qvm.qAlloc_many(n_qubit)
  qv_a = qvm.qAlloc_many(n_ancilla)
  qv = qv_w + qv_a

  lmbd = np.abs(coeffs).sum()
  probs = np.abs(coeffs) / lmbd
  amplitude = np.sqrt(probs)
  if DEBUG:
    print('lmbd:', lmbd)
    print('probs:', probs, 'sum=', probs.sum())
    print('amplitude:', amplitude)

  if not 'use Encode':
    encoder = Encode()
    encoder.amplitude_encode(qv_a, amplitude)
    PREP = encoder.get_circuit()
  else:
    PREP = amplitude_encode(qv_a, amplitude)
  PREP_dagger = PREP.dagger()

  qprog = QProg() << PREP
  probs = qvm.prob_run_list(qprog, qv_a)
  if DEBUG: print('probs (PREP):', probs)

  SEL = QCircuit()
  for i in range(len(terms)):
    r = i
    cond = QCircuit()
    for j in range(n_ancilla):
      if r & 1 == 0: cond << X(qv_a[j])
      r >>= 1
    cond_dagger = cond.dagger()
    term = terms[i]
    cu = QCircuit()
    j = 0
    sym = term[j]
    if sym == 'I': cu << I(qv_w[j]).control(qv_a)
    if sym == 'X': cu << X(qv_w[j]).control(qv_a)
    if sym == 'Z': cu << Z(qv_w[j]).control(qv_a)
    SEL << cond << cu << cond_dagger << BARRIER(qv)

  qcir = PREP << BARRIER(qv) << SEL << PREP_dagger
  print(draw_qprog(qcir, line_length=QCIR_LEN))
  if QCIR_SAVE: draw_qprog(qcir, line_length=QCIR_LEN, output='pic', filename='qcir-LCU.png')
  if DEBUG: show_circuit_matrix_UA(A, qvm, qv, qcir, lmbd)

print('[block_encoding_LCU]')
block_encoding_LCU(A_LCU)


def block_encoding_ARCSIN(A:ndarray, tol:float=1e-5) -> ndarray:
  ''' arxiv:2402.17529, the basis QUERY-ORACLE framework of FABLE '''
  N_ROW, N_COL = A.shape
  nq_row = int(np.ceil(np.log2(N_ROW)))
  nq_col = int(np.ceil(np.log2(N_COL)))
  n_qubit = max(nq_row, nq_col)
  n_qubit_ex = 1 + n_qubit * 2
  if DEBUG:
    print('n_qubit:', n_qubit)
    print('n_qubit_ex:', n_qubit_ex)
  lmbd = 2**n_qubit

  qvm = CPUQVM()
  qvm.init_qvm()
  # |c0,...,cn;r0,...,rm;a>
  qv = qvm.qAlloc_many(n_qubit_ex)
  # Fig. 3 from arxiv:2402.17529, mind the fucking order!!
  qv_col = qv[:n_qubit]
  qv_row = qv[n_qubit:-1]
  q_anc = qv[-1]
  qv_mctrl = qv[:-1]
  if DEBUG:
    print('len(qv_row):', len(qv_row))
    print('len(qv_col):', len(qv_col))
    print('len(qv_mctrl):', len(qv_mctrl))

  # Eq. 15 & 16 from arxiv:2205.00081, the basic "matrix query oracle" in FABLE essay
  # a_{i,j} = |a_{i,j}| exp(i * \alpha_{i,j})
  # \theta_{i,j} = arccos(|a_{i,j}|)
  # \phi_{i,j} = -\alpha_{i,j}
  # Eq. 6 from arxiv:2402.17529, here we use ARCSIN method for the real parts
  thetas = np.arcsin(np.abs(A))
  phis = np.angle(A)
  if DEBUG:
    print('thetas:', thetas)
    print('phis:', phis)

  qcir = QCircuit()
  for q in qv_row: qcir << H(q)
  qcir << BARRIER(qv)
  for i in range(N_ROW):
    for j in range(N_COL):
      t = 2 * thetas[i, j]
      p = 2 * phis  [i, j]
      if abs(t) < tol and abs(p) < tol: continue

      # cond
      cond = QCircuit()
      loc = i
      for q_r in qv_row:
        if loc & 1 == 0: cond << X(q_r)
        loc >>= 1
      loc = j
      for q_c in qv_col:
        if loc & 1 == 0: cond << X(q_c)
        loc >>= 1
      # mctrl-rot
      mcrot_circ = QCircuit()
      if abs(t) > tol: mcrot_circ << RY(q_anc, t).control(qv_mctrl)
      if abs(p) > tol: mcrot_circ << RZ(q_anc, p).control(qv_mctrl)
      # cond (uncompute)
      qcir << cond << mcrot_circ << cond << BARRIER(qv)
  for q_r, q_c in zip(qv_row, qv_col):
    qcir << SWAP(q_r, q_c)
  qcir << X(q_anc)    # ARCSIN flip ancilla
  qcir << BARRIER(qv)
  for q in qv_row: qcir << H(q)
  print(draw_qprog(qcir, line_length=QCIR_LEN))
  if QCIR_SAVE: draw_qprog(qcir, line_length=QCIR_LEN, output='pic', filename='qcir-ARCSIN.png')
  if DEBUG: show_circuit_matrix_UA(A, qvm, qv, qcir, lmbd)

print('[block_encoding_ARCSIN]')
block_encoding_ARCSIN(A_complex)


def block_encoding_FABLE_compute_theta(alpha:ndarray) -> ndarray:
  def _matrix_M_entry(row, col):
    # (col >> 1) ^ col is the Gray code of col
    b_and_g = row & ((col >> 1) ^ col)
    sum_of_ones = 0
    while b_and_g > 0:
      if b_and_g & 0b1:
        sum_of_ones += 1
      b_and_g = b_and_g >> 1
    return (-1) ** sum_of_ones

  ln = alpha.shape[-1]
  k = np.log2(ln)
  M_trans = np.zeros(shape=(ln, ln))
  for i in range(len(M_trans)):
    for j in range(len(M_trans[0])):
      M_trans[i, j] = _matrix_M_entry(j, i)
  if DEBUG: print('M_trans:', M_trans)
  theta = np.transpose(np.dot(M_trans, np.transpose(alpha)))
  return theta / 2**k

def block_encoding_FABLE_gray_code(rank:int) -> List[str]:
  def gray_code_recurse(g, rank):
    k = len(g)
    if rank <= 0: return

    for i in range(k - 1, -1, -1):
      char = "1" + g[i]
      g.append(char)
    for i in range(k - 1, -1, -1):
      g[i] = "0" + g[i]
    gray_code_recurse(g, rank - 1)

  g = ["0", "1"]
  gray_code_recurse(g, rank - 1)
  return g

def block_encoding_FABLE(A:ndarray, tol:float=1e-8) -> ndarray:
  n_qubit = int(np.ceil(np.log2(A.shape[0])))
  n_qubit_ex = 1 + n_qubit * 2
  if DEBUG:
    print('n_qubit:', n_qubit)
    print('n_qubit_ex:', n_qubit_ex)

  qvm = CPUQVM()
  qvm.init_qvm()
  qv = qvm.qAlloc_many(n_qubit_ex)
  # |a,i,j(working)>
  wires = list(reversed(range(n_qubit_ex)))
  ancilla = wires[0]
  wires_i = wires[1 : 1 + len(wires) // 2][::-1]
  wires_j = wires[1 + len(wires) // 2 : len(wires)][::-1]
  if DEBUG:
    print('wires:', wires)
    print('ancilla:', ancilla)
    print('wires_i:', wires_i)
    print('wires_j:', wires_j)
  
  code = block_encoding_FABLE_gray_code(2 * n_qubit)
  if DEBUG: print('code:', code)
  n_sel = len(code)
  ctrl_wires = [
    int(np.log2(int(code[i], 2) ^ int(code[(i + 1) % n_sel], 2)))
    for i in range(n_sel)
  ]
  wire_map = dict(enumerate(wires_j + wires_i))
  if DEBUG:
    print('ctrl_wires:', ctrl_wires)
    print('wire_map:', wire_map)

  alphas = np.arccos(A).flatten()
  thetas = block_encoding_FABLE_compute_theta(alphas)
  if DEBUG:
    print('alphas:', alphas)
    print('thetas:', thetas)
  lmbd = 2**n_qubit

  qcir = QCircuit()
  for i in wires_i: qcir << H(qv[i])
  qcir << BARRIER(qv)
  nots = set()
  for theta, ctrl_idx in zip(thetas, ctrl_wires):
    if abs(2 * theta) > tol:
      for ctrl_wire in nots:
        qcir << CNOT(qv[ctrl_wire], qv[ancilla])
      qcir << RY(qv[ancilla], 2 * theta)
      nots.clear()
    if wire_map[ctrl_idx] in nots:
      nots.remove(wire_map[ctrl_idx])
    else:
      nots.add(wire_map[ctrl_idx])
  for ctrl_wire in nots:
    qcir << CNOT(qv[ctrl_wire], qv[ancilla])
  qcir << BARRIER(qv)
  for i, j in zip(wires_i, wires_j):
    qcir << SWAP(qv[i], qv[j])
  qcir << BARRIER(qv)
  for i in wires_i: qcir << H(qv[i])
  print(draw_qprog(qcir, line_length=QCIR_LEN))
  if QCIR_SAVE: draw_qprog(qcir, line_length=QCIR_LEN, output='pic', filename='qcir-FABLE.png')
  if DEBUG: show_circuit_matrix_UA(A, qvm, qv, qcir, lmbd)

print('[block_encoding_FABLE]')
block_encoding_FABLE(A_real)

File: src/block_encoding.cpp
#include "Core/Utilities/QPandaNamespace.h"
#include "Core/Utilities/QProgInfo/QCircuitInfo.h"
#include "Components/Operator/PauliOperator.h"
#include "QAlg/Base_QCircuit/AmplitudeEncode.h"
#include "block_encoding.h"
#include "utils.h"

bool check_block_encoding(block_encoding_res &res, MatrixXcd &A, float eps=1e-5) {
  auto N = A.rows(), M = A.cols();
  auto block = res.unitary.block(0, 0, N, M) / res.subfactor;
  return (A - block).cwiseAbs().maxCoeff() < eps;
}


// Block-Encoding interface
MatrixXcd block_encoding(MatrixXcd H, BlockEncodingMethod method=BlockEncodingMethod::ARCSIN) {
  if (!is_shape_pow2(H)) {  // auto expand shape
    int N = H.rows(), M = H.cols();
    int maxN = N > M ? N : M;
    maxN = int(pow(ceil(log2(maxN)), 2));
    MatrixXcd H_ex = MatrixXcd::Zero(maxN, maxN);
    H_ex.block(0, 0, N, M) = H;
    H = H_ex;
  }
  block_encoding_res res;
  switch (method) {
    case BlockEncodingMethod::QSVT:   res = block_encoding_QSVT  (H); break;
    case BlockEncodingMethod::QSVT0:  res = block_encoding_QSVT0 (H); break;
    case BlockEncodingMethod::LCU:    res = block_encoding_LCU   (H); break;
    case BlockEncodingMethod::ARCSIN: res = block_encoding_ARCSIN(H); break;
    case BlockEncodingMethod::FABLE:  res = block_encoding_FABLE (H); break;
    default: throw invalid_argument("invalid method");
  }
  return res.unitary;
}


// Block-Encoding via QSVT-like direct construction from arXiv:2203.10236 Eq. 3.4
// Accepting arbitary square or non-sqaure matrix
block_encoding_res block_encoding_QSVT(MatrixXcd A) {
  if (spectral_norm(A) > 1) {
    cout << "spectral_norm: " << spectral_norm(A) << endl;
    throw domain_error("A must satisfy ||A||2 <= 1");
  }

  // https://pennylane.ai/qml/demos/tutorial_intro_qsvt/
  int N = A.rows(), M = A.cols();
  MatrixXcd U_A(N + M, N + M);
  U_A.block(0, 0, N, M) = A;
  U_A.block(0, M, N, N) = sqrt(MatrixXcd::Identity(N, N) - A * A.adjoint());
  U_A.block(N, 0, M, M) = sqrt(MatrixXcd::Identity(M, M) - A.adjoint() * A);
  U_A.block(N, M, M, N) = -A.adjoint();
  return block_encoding_res(U_A, 1);
}


// Block-Encoding via QSVT-like direct construction for 2x2 real symmetric matrix special case, refer to arXiv:2203.10236 Eq. 3.6
// Accepting 2x2 real symmetric matrix
block_encoding_res block_encoding_QSVT0(MatrixXcd A) {
  if (!is_square(A)) throw domain_error("A is not square");
  if (A.rows() != 2) throw domain_error("A is not 2x2 shape");
  if (!is_real(A)) throw domain_error("A is not real");
  if (spectral_norm(A) > 1) throw domain_error("A must satisfy ||A||2 <= 1");
  auto p = A(0, 0), q = A(0, 1),
       r = A(1, 0), s = A(1, 1);
  const float eps = 1e-8;
  if (abs(p - s) > eps) throw domain_error("A is not symmetric along main diagonal");
  if (abs(q - r) > eps) throw domain_error("A is not symmetric along sub diagonal");

  auto a1 = p.real(), a2 = q.real();
  auto b1 = sqrt(1 - pow(a1, 2)), b2 = sqrt(1 - pow(a2, 2));
  MatrixXcd U_a = MatrixXcd(4, 4);
  U_a << a1, a2, a1, -a2,
         a2, a1, -a2, a1,
         a1, -a2, a1, a2,
         -a2, a1, a2, a1;
  MatrixXcd U_b = MatrixXcd(4, 4);
  U_b << b1, b2, b1, -b2,
         b2, b1, -b2, b1,
         b1, -b2, b1, b2,
         -b2, b1, b2, b1;
  MatrixXcd U_A(8, 8);
  U_A << U_a, -U_b,
         U_b, U_a;
  return block_encoding_res(U_A / 2, 1 / 2);
}


// Block-Encoding via linear combination of unitaries method from arXiv:1501.01715
// Accepting square matrix, decomposable to a linear combination of unitaries
block_encoding_res block_encoding_LCU(MatrixXcd A, float eps=1e-8) {
  if (A.rows() != 2) throw domain_error("A is not 2x2 shape, currently only support 2x2 matrix");
  if (!is_shape_pow2(A)) throw domain_error("A shape is not power of 2");
  if (!is_real(A)) throw domain_error("A is not real, currently only support real matrix");

  // matrix to LCU
  int N = A.rows();
  EigenMatrixX matrix = A.real();
  PauliOperator ops = matrix_decompose_hamiltonian(matrix);
  vector<float> coeffs;
  vector<string> terms;
  for (auto item : ops.data()) {
    double coeff = item.second.real();
    if (abs(coeff) < eps) continue;
    coeffs.push_back(coeff);
    terms.push_back(item.first.first.size() == 0 ? "I0" : item.first.second);
  }

  // LCU to QCircuit
  // 1. decide system size
  size_t n_qubit = ceil(log2(N));
  size_t n_unitary = coeffs.size();
  size_t n_ancilla = ceil(log2(n_unitary));
  size_t n_qubit_ex = n_qubit + n_ancilla;
  CPUQVM qvm;
  qvm.setConfigure({n_qubit_ex, n_qubit_ex});
  qvm.init();
  // |...,a1,a0,...,q1,q0>, mind the fucking order!! :(
  QVec qv_w = qvm.qAllocMany(n_qubit);
  QVec qv_a = qvm.qAllocMany(n_ancilla);
  QVec qv = qv_w + qv_a;
  // 2. normalize coeffs & prepare ancilla state (PREP)
  float lmbd = 0.0;
  for (auto coeff : coeffs) lmbd += abs(coeff);
  // special case: only one unitary
  if (n_unitary == 1) {
    return block_encoding_res(A / lmbd, lmbd);
  }
  vector<double> probs;
  for (auto coeff : coeffs) probs.push_back(abs(coeff) / lmbd);
  vector<double> amplitude;
  for (auto prob : probs) amplitude.push_back(sqrt(prob));
  QCircuit PREP = amplitude_encode(qv_a, amplitude, false);
  // 3. make selection on unitaries (SEL)
  QCircuit SEL;
  for (int i = 0; i < terms.size(); i++) {
    // condition
    int r = i;
    QCircuit cond;
    for (int j = 0; j < n_ancilla; j++) {
      if (r % 2 == 0) cond << X(qv_a[j]);
      r /= 2;
    }
    // unitary
    string term = terms[i];
    QCircuit cu;
    int j = 0;    // FIXME: current only support 1-qubit :(
    auto sym = term.size() > 0 ? term[j] : 'I';
    switch (sym) {
      case 'I': cu << I(qv_w[j]).control(qv_a); break;
      case 'X': cu << X(qv_w[j]).control(qv_a); break;
      //case 'Y': cu << Y(qv_w[j]).control(qv_a); break;    // QPanda not support
      case 'Z': cu << Z(qv_w[j]).control(qv_a); break;
    }
    SEL << createEmptyCircuit() << cond << cu << cond.dagger() << BARRIER(qv);
  }
  // 4. build total circuit
  QCircuit qcir = createEmptyCircuit() << PREP << BARRIER(qv) << SEL << PREP.dagger();

  // QCircuit to matrix
  QProg prog = createEmptyQProg() << qcir;
  QStat mat_flat = getCircuitMatrix(prog, true);
  qvm.finalize();
  int N_ex = pow(2, n_qubit_ex);
  MatrixXcd U_A = Map<MatrixXcd>(mat_flat.data(), N_ex, N_ex);

  return block_encoding_res(U_A, 1 / lmbd, qcir);
}


// Block-Encoding via ARCSIN method from arXiv:2402.17529
// Accepting d-sparse square matrix with |a_ij| <= 1
block_encoding_res block_encoding_ARCSIN(MatrixXcd A, float eps=1e-8) {
  if (!is_shape_pow2(A)) throw domain_error("A shape is not power of 2");
  if (!is_elem_norm(A)) throw domain_error("A elememnts must st. |a_ij| <= 1");

  // build circuit
  // 1. decide system size
  int N = A.rows(), M = A.cols();
  size_t n_qubit = ceil(log2(N));
  size_t n_qubit_ex = 1 + n_qubit * 2;
  CPUQVM qvm;
  qvm.setConfigure({n_qubit_ex, n_qubit_ex});
  qvm.init();
  // |c0,...,cn;r0,...,rm;a>, mind the fucking order!! :(
  QVec qv = qvm.qAllocMany(n_qubit_ex);
  QVec qv_col = QVec(qv.begin(), qv.begin() + n_qubit);
  QVec qv_row = QVec(qv.begin() + n_qubit, qv.begin() + 2 * n_qubit);
  QVec qv_mctrl = QVec(qv.begin(), qv.end() - 1);
  Qubit* q_anc = qv[qv.size() - 1];
  // 2. make oracle O_A
  MatrixXd thetas = A.array().abs().asin().matrix();
  MatrixXd phis = A.array().arg().matrix();
  QCircuit O_A;
  for (int i = 0; i < N; i++) {
    for (int j = 0; j < M; j++) {
      float t = 2 * thetas(i, j);
      float p = 2 * phis(i, j);
      if (abs(t) < eps && abs(p) < eps) continue;

      // cond
      QCircuit cond;
      int loc = i;
      for (auto q_r : qv_row) {
        if (loc % 2 == 0) cond << X(q_r);
        loc /= 2;
      }
      loc = j;
      for (auto q_c : qv_col) {
        if (loc % 2 == 0) cond << X(q_c);
        loc /= 2;
      }
      // mctrl-rot
      QCircuit mcrot_circ;
      if (abs(t) > eps) mcrot_circ << RY(q_anc, t).control(qv_mctrl);
      if (abs(p) > eps) mcrot_circ << RZ(q_anc, p).control(qv_mctrl);
      // combine one layer
      O_A << cond << mcrot_circ << cond << BARRIER(qv);
    }
  }
  // 3. make H, SWAP and flip-X
  QCircuit H_circ;
  for (auto q : qv_row) H_circ << H(q);
  QCircuit SWAP_circ;
  for (int k = 0; k < qv_row.size(); k++) SWAP_circ << SWAP(qv_row[k], qv_col[k]);
  SWAP_circ << X(q_anc);
  // 4. build total circuit
  QCircuit qcir = createEmptyCircuit() << H_circ << BARRIER(qv) << O_A << BARRIER(qv) << SWAP_circ << BARRIER(qv) << H_circ;

  // QCircuit to matrix: 穷举制备每个基态，逐列投影出线路所对应的矩阵
  // XXX: getCircuitMatrix() does NOT work for this, IDK why :(
  int N_ex = pow(2, n_qubit_ex);
  MatrixXcd U_A(N_ex, N_ex);
  for (int i = 0; i < N_ex; i++) {
    int r = i;
    QCircuit cond;
    for (int j = 0; j < n_qubit_ex; j++) {
      if (r % 2 == 1) cond << X(qv[j]);
      r /= 2;
    }
    QProg qprog = createEmptyQProg() << cond << qcir;
    qvm.directlyRun(qprog);
    QStat qs = qvm.getQState();
    for (int j = 0; j < N_ex; j++) 
      U_A(j, i) = qs[j];
  }
  qvm.finalize();

  return block_encoding_res(U_A, 1 / float(N), qcir);
}


// Block-Encoding via FABLE method from arXiv:2205.00081
// Accepting d-sparse square matrix with |a_ij| <= 1
// @impl translated from https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/templates/subroutines/fable.py
VectorXd block_encoding_FABLE_compute_theta(VectorXd alpha) {
  int N = alpha.size();
  MatrixXd M_trans = MatrixXd::Zero(N, N);
  for (int i = 0; i < M_trans.rows(); i++)
    for (int j = 0; j < M_trans.cols(); j++) {
      int row = j, col = i;     // XXX: the order seems transposed, IDK why :(
      int b_and_g = row & ((col >> 1) ^ col);
      int sum_of_ones = 0;
      while (b_and_g > 0) {
        if (b_and_g % 2 == 1)
          sum_of_ones += 1;
        b_and_g /= 2;
      }
      M_trans(i, j) = sum_of_ones % 2 ? -1 : 1;
    }
  return M_trans * alpha / N;
}

void block_encoding_FABLE_gray_code_recursive(vector<string> &g, int rank) {
  if (rank <= 0) return;
  size_t k = g.size();
  for (int i = k - 1;i >= 0; i--)
    g.push_back("1" + g[i]);
  for (int i = k - 1;i >= 0; i--)
    g[i] = "0" + g[i];
  block_encoding_FABLE_gray_code_recursive(g, rank - 1);
}

vector<int> block_encoding_FABLE_gray_code(int rank) {
  vector<string> g = {"0", "1"};
  block_encoding_FABLE_gray_code_recursive(g, rank - 1);
  vector<int> code;
  for (auto c : g) code.push_back(stoi(c.c_str(), nullptr, 2));
  return code;
}

block_encoding_res block_encoding_FABLE(MatrixXcd A, float eps=1e-8) {
  if (!is_shape_pow2(A)) throw domain_error("A shape is not power of 2");
  if (!is_real(A)) throw domain_error("A is not real");
  if (!is_elem_norm(A)) throw domain_error("A elememnts must st. |a_ij| <= 1");

  // build circuit
  // 1. decide system size
  int N = A.rows();
  size_t n_qubit = ceil(log2(N));
  size_t n_qubit_ex = 1 + n_qubit * 2;
  CPUQVM qvm;
  qvm.setConfigure({n_qubit_ex, n_qubit_ex});
  qvm.init();
  // |a,i,j(working)>, mind the fucking order!! :(
  QVec qv = qvm.qAllocMany(n_qubit_ex);
  vector<int> wires, wires_i, wires_j;
  for (int i = n_qubit_ex - 1; i >= 0; i--) wires.push_back(i);
  int n_wire = wires.size();
  int anc = wires[0];
  for (int k = n_wire / 2; k >= 1; k--) wires_i.push_back(wires[k]);
  for (int k = n_wire - 1; k >= n_wire / 2 + 1; k--) wires_j.push_back(wires[k]);
  vector<int> code = block_encoding_FABLE_gray_code(2 * n_qubit);
  int n_sel = code.size();
  vector<int> ctrl_wires;
  for (int i = 0; i < n_sel; i++) {
    int idx = log2(code[i] ^ code[(i + 1) % n_sel]);
    ctrl_wires.push_back(idx);
  }
  map<int, int> wire_map; {
    int idx = 0;  // make it tmp var
    for (auto k : wires_j) { wire_map[idx] = k; idx++; }
    for (auto k : wires_i) { wire_map[idx] = k; idx++; }
  }
  // 2. make oracle O_A
  Matrix<double, Dynamic, Dynamic, RowMajor> A_rm(A.real());
  Map<RowVectorXd> A_flat(A_rm.data(), A_rm.size());
  VectorXd alphas = A_flat.real().array().acos();
  VectorXd thetas = block_encoding_FABLE_compute_theta(alphas);
  QCircuit O_A;
  set<int> nots;
  for (int k = 0; k < thetas.size(); k++) {
    float theta = thetas(k);
    int ctrl_idx = ctrl_wires[k];
    if (abs(2 * theta) > eps) {
      for (auto ctrl_wire : nots)
        O_A << CNOT(qv[ctrl_wire], qv[anc]);
      O_A << RY(qv[anc], 2 * theta);
      nots.clear();
    }
    if (nots.find(wire_map[ctrl_idx]) != nots.end())
      nots.erase(wire_map[ctrl_idx]);
    else
      nots.insert(wire_map[ctrl_idx]);
  }
  for (auto ctrl_wire : nots)
    O_A << CNOT(qv[ctrl_wire], qv[anc]);
  // 3. make H and SWAP
  QCircuit H_circ;
  for (auto i : wires_i) H_circ << H(qv[i]);
  QCircuit SWAP_circ;
  for (int k = 0; k < wires_i.size(); k++) SWAP_circ << SWAP(qv[wires_i[k]], qv[wires_j[k]]);
  // 4. build total circuit
  QCircuit qcir = createEmptyCircuit() << H_circ << BARRIER(qv) << O_A << BARRIER(qv) << SWAP_circ << BARRIER(qv) << H_circ;

  // QCircuit to matrix: 穷举制备每个基态，逐列投影出线路所对应的矩阵
  // XXX: getCircuitMatrix() does NOT work for this, IDK why :(
  int N_ex = pow(2, n_qubit_ex);
  MatrixXcd U_A(N_ex, N_ex);
  for (int i = 0; i < N_ex; i++) {
    int r = i;
    QCircuit cond;
    for (int j = 0; j < n_qubit_ex; j++) {
      if (r % 2 == 1) cond << X(qv[j]);
      r /= 2;
    }
    QProg qprog = createEmptyQProg() << cond << qcir;
    qvm.directlyRun(qprog);
    QStat qs = qvm.getQState();
    for (int j = 0; j < N_ex; j++) 
      U_A(j, i) = qs[j];
  }
  qvm.finalize();

  return block_encoding_res(U_A, 1 / float(N), qcir);
}


// ↓↓ keep signature for the contest solution
MatrixXcd block_encoding_method(MatrixXcd H) {
  // 由于测试数据不保证谱范数、系数值域等等条件，我们使用应用范围最广的 QSVT 作为最终提交解答；
  // 即使它不存在理论意义上复杂度低的门线路分解方案 :(
  H = rescale_if_necessary(H);
  return block_encoding(H, BlockEncodingMethod::QSVT);
}
